### ISSUE TITLE
Invalid type in except clause

### ISSUE THREAD

**Issue by tkukushkin** [microsoft/pyright#2648]
**Describe the bug**
Pyright shows type of variable in `except` clause as It's like there's no `try` body and variable type annotation.

**To Reproduce**
```python
from typing import Optional

x: Optional[str] = None
try:
    x = '10'
except Exception:
    reveal_type(x)  # info: Type of "x" is "None"
    if x: 
        reveal_type(x)  # info: Type of "x" is "Never"
```

**Expected behavior**
Type of `x` expected to be `Optional[str]` in except clause. This is how mypy works ([playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=b5ddadd5e67839ec717c03054c9ee8d0)).

**VS Code extension or command-line**
```bash
❯ pyright --version
pyright 1.1.191
```

**Comment by erictraut** [comment]
How is it possible for an exception to be generated after the `x = '10'` assignment? 

If I add a statement (e.g. a call) after `x = '10'` that could potentially generate an exception, then pyright evaluates the types you're expecting to see.

```python
def possible_exception() -> int:
    ...

x: Optional[str] = None
try:
    x = "10"
    possible_exception()
except Exception:
    reveal_type(x)  # info: Type of "x" is "Literal[10] | None"
    if x:
        reveal_type(x)  # info: Type of "x" is "Literal[10]"
```

**Comment by tkukushkin** [comment]
Oops, sorry, all because I simplified the example.

My original code uses `with` statement.

```python
from typing import Optional, ContextManager

def create_context() -> ContextManager[str]: ...
def possible_exception() -> None: ...

x: Optional[str] = None
ctx: Optional[str] = None
try:
    with create_context() as ctx:
        x = '10'
        possible_exception()
except Exception:
    reveal_type(x)  # info: Type of "x" is "None"
    reveal_type(ctx)  # info: Type of "x" is "None"
```

**Comment by erictraut** [comment]
Thanks for the additional context. 

Nested try/except statements are challenging for code flow analysis. The code in your sample is an even more challenging case because it involves a context manager (which potentially involves an implied try/except) nested within a try/except statement. This is more challenging because the code flow graph (which is needed to perform type analysis) needs to accommodate context managers that swallow exceptions and those that don't. 

As a fix for this issue, I've added an implicit vertex in the code flow graph from the _end_ of the context manager block to the except blocks. This doesn't exactly model the runtime code flow paths, but it's "more correct" that the previous approach, and it results in the correct type evaluation in your sample.

This change will be included in the next release.

**Comment by tkukushkin** [comment]
Thank you for such quick fixes!

**Comment by erictraut** [comment]
This is included in pyright 1.1.193, which I just published. It will also be included in the next release of pylance.

**Comment by tkukushkin** [comment]
@erictraut found that the fix did not help for the case when there is a `return` inside the `with` statement.

```python
from typing import Optional, ContextManager

def create_context() -> ContextManager[str]: ...
def possible_exception() -> None: ...

def main() -> None:
    ctx: Optional[str] = None
    try:
        with create_context() as ctx:
            possible_exception()
            return
    except Exception:
        reveal_type(ctx)  #  information: Type of "ctx" is "None"
```
